src/lib/ast.rs

```diff
@@ -1,11 +1,48 @@
+use std::ops::Range;
+
 use serde::{Deserialize, Serialize};
 
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Loc<X>(pub X, pub Source);
```

Comment:
Does it help in any way to absorb the `Box<_>` around the `X` into this type definition?

That way, we avoid having to think about the Box and the Loc separately, and they combine in the AST structure.

Reply:
I originally included `Box<_>` in the definition, but I ran into edge cases where I just wanted the source information without storing everything on the heap (specifically for the lexer). This could also give us room for optimizations down the line in the interpreter.